node/network/statement-distribution/src/lib.rs

```diff
@@ -907,17 +924,39 @@ async fn circulate_statement(
 	ctx: &mut impl SubsystemContext,
 	relay_parent: Hash,
 	stored: &StoredStatement,
+	mut priority_peers: Vec<PeerId>,
 ) -> Vec<PeerId> {
 	let fingerprint = stored.fingerprint();
 
-	let peers_to_send: Vec<PeerId> = peers.iter().filter_map(|(peer, data)| {
+	let mut peers_to_send: Vec<PeerId> = peers.iter().filter_map(|(peer, data)| {
 		if data.can_send(&relay_parent, &fingerprint) {
 			Some(peer.clone())
 		} else {
 			None
 		}
 	}).collect();
-	let peers_to_send = util::choose_random_sqrt_subset(peers_to_send, MIN_GOSSIP_PEERS);
+
+	let good_peers: HashSet<&PeerId> = peers_to_send.iter().collect();
+	// Only take priority peers we can send data to:
+	priority_peers.retain(|p| good_peers.contains(p));
+
+	// Avoid duplicates:
+	let priority_set: HashSet<&PeerId> = priority_peers.iter().collect();
+	peers_to_send.retain(|p| !priority_set.contains(p));
+
+	let mut peers_to_send =
+		util::choose_random_sqrt_subset(peers_to_send, MIN_GOSSIP_PEERS);
+	// We don't want to use less peers, than we would without any priority peers:
+	let min_size = std::cmp::max(peers_to_send.len(), MIN_GOSSIP_PEERS);
+	// Make set full:
+	let needed_peers = min_size as i64 - priority_peers.len() as i64;
```

Comment:
Could `priority_peers.len()` not be greater than `min_size` here?

Reply:
It can (that's even tested in the test suite), but is fine. The if below will do the right thing in that case.

Reply:
I see - the cast to `i64` prevents underflow